### PR TITLE
Blynk option and email notification

### PR DESCRIPTION
**Pull request by ldab**
Merged commit into master. Email option for low battery.

#259

Also option to trigger calibration and configuration via App Terminal possible if we want. #TODO

### CODE REVIEW

**Comment by Grandement** [comment]
I tried it and works well. It would be better to extend the docs with the virtual pin definition. Because I would like to reuse my token instead of copy your project with a new token. Therefore it was additional effort to look in the code for the correct pins.
The option to configure it via the terminal would be really help. This did not work for me but you already mentioned it (#TODO) 

Thanks for the awesome work.

**Comment by ldab** [comment]
> 
> 
> I tried it and works well. It would be better to extend the docs with the virtual pin definition. Because I would like to reuse my token instead of copy your project with a new token. Therefore it was additional effort to look in the code for the correct pins.
> The option to configure it via the terminal would be really helpful. This did not work for me but you already mentioned it (#TODO)
> 
> Thanks for the awesome work.

Thanks for the input, once the pull request (if) is approved, I will update the docs accordingly.

**Comment by ldab** [comment]
Have rebased, but don't have the HW to test it crashes without the sensors, are you able to try?

**Comment by ldab** [comment]
Tested and fixed, all good :)

**Comment by universam1** [comment]
@ldab would you mind to solve the merge conflicts?

**Comment by ldab** [comment]
@universam1 done

File: pio/lib/Sender/Sender.cpp [hunk 1]
```diff
@@ -485,3 +487,237 @@ bool SenderClass::sendTCONTROL(String server, uint16_t port)
...
+    if (_validflag != RTCVALIDFLAG)
+    {
+      Blynk.email(Subject, Content);
+      _validflag = RTCVALIDFLAG;
+      ESP.rtcUserMemoryWrite(RTCEMAILADDR, &_validflag, sizeof(_validflag));
+      delay(200);
+    }
+}
+
+uint8_t SenderClass::blynkTerminal(String token)
```
Comment (universam1) [inline pio/lib/Sender/Sender.cpp]: I am not in favor for another configuration procedure besides of Wifi Manager UI. It is narrow minded, possibly fitting to Blynk but not at all for general purpose. It’s a scope creep

Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: `::blynkTerminal` ? It's not fully implemented yet on this PR... The idea behind it is to trigger the configuration without opening the container.

We can simply remove it without affecting the current functionality, please let me know.

Comment (universam1) [reply pio/lib/Sender/Sender.cpp]: Yes please remove it then

File: pio/lib/Sender/Sender.cpp [hunk 2]
```diff
@@ -485,3 +487,237 @@ bool SenderClass::sendTCONTROL(String server, uint16_t port)
...
+    else {
+        CONSOLELN(F("\nFailed to connect to Blynk, going to sleep"));
+        return false;
+    }
+    
+    delay(300);     //delay to allow last value to be sent;
+    return true;
+}
+
+bool SenderClass::sendBlynkHTTP(String token)
```
Comment (universam1) [inline pio/lib/Sender/Sender.cpp]: If this http method is not used, can we remove the dead code?

File: pio/lib/Sender/Sender.cpp [hunk 3]
```diff
@@ -485,3 +487,237 @@ bool SenderClass::sendTCONTROL(String server, uint16_t port)
...
+        else{
+            CONSOLELN(F("\nERROR Sender: couldnt connect"));
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void SenderClass::sendBlynkEmail(String Subject, String Content)
```
Comment (universam1) [inline pio/lib/Sender/Sender.cpp]: Not certain about this alerting method, normally the server side is responsible for alerts not the client! Wonder what can go wrong and delay which eventually kills the battery fully

File: pio/lib/Sender/Sender.cpp [hunk 4]
```diff
@@ -485,3 +487,237 @@ bool SenderClass::sendTCONTROL(String server, uint16_t port)
...
+            Blynk.virtualWrite(_pin, _value);
+        }
+    }
+
+    else {
+        CONSOLELN(F("\nFailed to connect to Blynk, going to sleep"));
+        return false;
+    }
+    
+    delay(300);     //delay to allow last value to be sent;
```
Comment (universam1) [inline pio/lib/Sender/Sender.cpp]: That’s obsolete, is been taken care of at the main caller

File: pio/lib/Sender/Sender.cpp (hunk 1 above)
Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: ok

File: pio/lib/Sender/Sender.cpp (hunk 3 above)
Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: ok

File: pio/lib/Sender/Sender.cpp (hunk 4 above)
Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: Not sure what is obsolete here, can you please clarify?

File: pio/lib/Sender/Sender.cpp (hunk 2 above)
Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: ok

File: pio/lib/Sender/Sender.cpp (hunk 3 above)
Comment (universam1) [reply pio/lib/Sender/Sender.cpp]: I’ve implemented a connection timeout on every call to prevent hanging connection attempts. For this blynk.email I doubt a bit this is super safe, especially the condition here where we are out of battery. Hence if there are no important reasons I’d consider the api as the only communication way without another sidechannel

File: pio/lib/Sender/Sender.cpp (hunk 4 above)
Comment (universam1) [reply pio/lib/Sender/Sender.cpp]: The delay here is covered already in the main routine, doesn’t need duplication

Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: This is required for Blynk for some reason, without it the last value is not sent. I tested with 100ms and it works. So I'm afraid we need to leave it there.

File: pio/lib/Sender/Sender.cpp (hunk 3 above)
Comment (ldab) [reply pio/lib/Sender/Sender.cpp]: The reason would be to warn the user about low battery, and it would only happen once. But as the brewering process goes, I would doubt someone would open the fermenter to charge the battery, so i will remove it.

**Comment by Topspliff** [comment]
Are we going to see this incorporated into a new release soon?
